### ISSUE TITLE
React-native won't compile because of the use of h3-js

### ISSUE THREAD

**Issue by johhnsmmith198** [uber/h3-js#117]
importing h3-js this way `import * as h3 from "h3-js";`

would trigger the error below:
```
[Sat May 08 2021 16:50:27.116]  ERROR    Error: Requiring module "node_modules/h3-js/dist/browser/h3-js.js", which threw an exception: ReferenceError: Can't find variable: document
[Sat May 08 2021 16:50:27.132]  ERROR    ReferenceError: Can't find variable: document
```

I am wondering if this could be easily fixed? I am reading that there is a node version of h3-js, how can I import it instead of the browser version?

I tried to only import the function I was using at the time but I am still getting the same error 
```
import {geoToH3} from 'h3-js';
```

I think the error is pointing to this code here in h3-js
```
  var readAsync;

  {
    if (document.currentScript) {
      scriptDirectory = document.currentScript.src;
    }

    if (scriptDirectory.indexOf("blob:") !== 0) {
      scriptDirectory = scriptDirectory.substr(0, scriptDirectory.lastIndexOf("/") + 1);
    } else {
      scriptDirectory = "";
    }
```

**Comment by nrabinowitz** [comment]
See https://github.com/uber/h3-js/issues/35 - this is a known issue. We're not currently targeting React Native; I'd like this to work, obviously, but the issue is deep in the emscripten internals and hard to remove. There's a [fork here](https://github.com/realPrimoh/h3-reactnative) that tries to address this, but I haven't had time to see if it can be fixed in the main library, and I don't have a great way to test the fix.

**Comment by johhnsmmith198** [comment]
A very simple way to fix it is to change the generated code in dist from:
```
  } else if (ENVIRONMENT_IS_WEB || ENVIRONMENT_IS_WORKER) {
    if (ENVIRONMENT_IS_WORKER) {
      scriptDirectory = self.location.href;
    } else if (document.currentScript) {
      scriptDirectory = document.currentScript.src;
    }
```
to 
```
  } else if (ENVIRONMENT_IS_WEB || ENVIRONMENT_IS_WORKER) {
    if (ENVIRONMENT_IS_WORKER) {
      scriptDirectory = self.location.href;
    } else if (ENVIRONMENT_IS_WEB && document.currentScript) {
      scriptDirectory = document.currentScript.src;
    }
```
Note the: `if (ENVIRONMENT_IS_WEB && document.currentScript) {` 

it seems that this is generated by `emscripten` => https://github.com/emscripten-core/emscripten/issues/6717

**Comment by padawannn** [comment]
> It looks like the bug on the Emscripten side ([emscripten-core/emscripten#14198](https://github.com/emscripten-core/emscripten/issues/14198)) was fixed as of November of last year. Would it be possible to cut a new h3-js release with the latest emscripten, which hopefully will make this library React Native compatible?

I'm trying to use h3-js in a web-worker and I have the same problem. It would be great if a new  h3-js release was published with the latest emscripten version

**Comment by corbt** [comment]
> I will note that the reason why emscripten was not updated for this project was due to performance regressions with asm.js

Oh interesting, I wasn't aware! Are there benchmarks for how large the impact would be on this library if it adopted the latest emscripten? Making this compatible with h3 would be really nice, but of course not at the cost of a significant performance regression for other users!

**Comment by dfellis** [comment]
It was a few years ago that we noticed the regression. I searched but couldn't find the information. As I recall most functions were 10-20% slower and some were more than 50% slower *but* since it has literally been years in the meantime, it's probably a good idea to try again and re-benchmark against the current version of emscripten.

**Comment by lofsigma** [comment]
I would also be interested in a version of h3-js that works on the server. 
I tried updating to the latest version of Emscripten  to at least get a benchmark but I couldn't get it to work.
- `MODULARIZE_INSTANCE=1` is deprecated so you have to use `MODULARIZE=1` instead, which returns a function which returns a promise that resolves to an instance.
- Also I think you need to explicitly export `_free` now.

**Comment by nrabinowitz** [comment]
> I would also be interested in a version of h3-js that works on the server. 

To be clear, `h3-js` works fine on a server running Node. It only has issues running in React Native.

**Comment by lofsigma** [comment]
Sorry I misspoke. It doesn't work with vercel's [edge runtime / middleware](https://edge-runtime.vercel.sh/). 
Same error as react native.
```
ReferenceError: document is not defined
```

**Comment by nrabinowitz** [comment]
Thanks for the clarification, that's good to know.

**Comment by cnrdh** [comment]
With Deno's recent support for npm packages, you may now use a regular import like
```js 
import * as h3 from "npm:h3-js@4.0.1";
```
Easy enough, but non-standard and won't help people on other runtimes (and not even Deno Deploy as the feature is experimental). It would be great if h3-js could simply get rid of the `document`-reference.

**Comment by kenjinp** [comment]
I've encountered this issue when trying to use h3 inside a web-worker as well

**Comment by nrabinowitz** [comment]
Related discussion, with benchmarks for the emscripten upgrade issue: https://github.com/uber/h3-js/issues/163#issuecomment-1380986297

**Comment by 4nthonylin** [comment]
This issue should be resolved with the most recent 4.1.0 release. I have been able to use the 4.1.0 dist directly in WebWorkers now greatly speeding up computations :)

**Comment by nrabinowitz** [comment]
Thanks @4nthonylin! Closing for now - there may be other blockers for React Native or unusual environments, but hopefully this is a reasonable fix for many of the affected situations.
